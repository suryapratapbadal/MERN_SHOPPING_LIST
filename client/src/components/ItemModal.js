import React, { Component } from 'react';
// import {
//     Button,
//     Modal,
//     ModalBody,
//     ModalHeader,
//     Form,
//     FormGroup,
//     Label,
//     Input
// } from 'reactstrap';

import { withStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Modal from '@material-ui/core/Modal';
import Button from '@material-ui/core/Button';
import AddIcon from '@material-ui/icons/Add';
import CssBaseline from '@material-ui/core/CssBaseline';
import FormControl from '@material-ui/core/FormControl';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';

import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { addItem } from '../actions/index';



export class ItemModal extends Component {
    constructor(props) {
        super(props);
        this.state = {
            modal: false,
            name: '',
        };
    }

    toggle = () => {
        this.setState({ modal: !this.state.modal });
    }

    onChange = event => {
        this.setState({ name: event.target.value });
    }

    onSubmit = event => {
        event.preventDefault();
        const newItem = {
            name: this.state.name,
        };


        this.props.addItem(newItem);
        this.toggle();
    }

    render() {

        const { classes } = this.props;

        return (
            <div className={classes.root}>
                <CssBaseline />
                <Button variant="contained" className={classes.button} onClick={this.toggle} disabled={this.props.loading}>
                    Add Item
                    <AddIcon className={classes.rightIcon} />
                </Button>
                <Modal
                    aria-labelledby="simple-modal-title"
                    aria-describedby="simple-modal-description"
                    open={this.state.modal}
                    onClose={this.toggle}
                >
                    <div className={classes.paper}>
                        <Typography variant="title" id="modal-title" className={classes.tittle}>
                            Add to Shopping List
                        </Typography>
                        <form className={classes.form} onSubmit={this.onSubmit}>
                            <FormControl margin="normal" required fullWidth>
                                <InputLabel htmlFor="item">Item Name</InputLabel>
                                <Input id="item" name="item" autoComplete="Add shooping item" autoFocus onChange={this.onChange} />
                            </FormControl>
                            <Button
                                type="submit"
                                fullWidth
                                variant="raised"
                                className={classes.submit}
                                disabled={this.state.name === ''}
                            >
                                Add Item
                            </Button>
                        </form>
                    </div>
                </Modal>
            </div>
        );
    }
}

const styles = theme => ({
    root: {
        marginTop: '2rem',
        marginBottom: '2rem'
    },
    rightIcon: {
        marginLeft: theme.spacing.unit,
    },
    paper: {
        // position: 'absolute',
        width: theme.spacing.unit * 50,
        backgroundColor: theme.palette.background.paper,
        boxShadow: theme.shadows[5],
        padding: theme.spacing.unit * 4,
        margin: 'auto'
    },
    button: {
        margin: theme.spacing.unit,
    },
    tittle: {
        textAlign: 'center',
        paddingBottom: '1rem',
        marginBottom: '1rem',
    },
    form: {
        width: '100%', // Fix IE11 issue.
        marginTop: theme.spacing.unit,
    },
    submit: {
        marginTop: theme.spacing.unit * 3,
    },
});

export default connect(state => {
    const loading = state.itemReducer.loading || false;
    return {
        loading
    }
}, dispatch => {
    return bindActionCreators({ addItem: addItem }, dispatch)
}
)(withStyles(styles)(ItemModal));