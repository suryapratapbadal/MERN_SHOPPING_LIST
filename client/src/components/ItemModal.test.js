import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
// import configureStore from 'redux-mock-store'; // Smart components

// Component to be tested
import { ItemModal } from './ItemModal';

// const mockStore = configureStore();
// const initialState = {
//   items: [],
//   loading: false
// }

// const store = mockStore(initialState);

// Component Renders
describe('<ItemModal />', () => {
  const mockaddItemfn = jest.fn();

  describe('render()', () => {
    test('renders the component', () => {
      const wrapper = shallow(<ItemModal addItem={mockaddItemfn} />);
      expect(toJson(wrapper)).toMatchSnapshot();

    });
  });
});


describe('<ItemModal />', () => {
  const mockaddItemfn = jest.fn();
  describe('onClick()', () => {
    test('successfully calls the onClick handler', () => {
      const wrapper = shallow(<ItemModal addItem={mockaddItemfn} />);
      const mockOnClick = jest.fn();
      const toggleState = wrapper.find('Modal').props().isOpen;
      wrapper.find('Button').at(0).simulate('click');

      expect(wrapper.find('Modal').props().isOpen).toEqual(!toggleState);
    });
  });

  describe('onChange()', () => {
    test('successfully calls the onChange handler', () => {
      const wrapper = shallow(<ItemModal addItem={mockaddItemfn} />);
      const mockOnChange = jest.fn();
      const input = wrapper.find('Input');

      input.props().onChange({
        target: {
          name: 'myName',
          value: 'myValue'
        }
      });
      // wrapper.find('Input').simulate('change', event);

      expect(wrapper.state('name')).toEqual('myValue');
    });

  });


  describe('onSubmit()', () => {
    test('successfully calls the onSubmit handler', () => {
      const wrapper = shallow(<ItemModal addItem={mockaddItemfn} />);
      const mockOnSubmitfn = jest.fn();
      const mockpreventDefaultfn = jest.fn()
      const form = wrapper.find('Form');
      const newItem = {
        name: 'Eggs'
      };


      wrapper.find('Form').simulate('submit', { preventDefault: () => { } });

      expect(mockaddItemfn).toHaveBeenCalledTimes(1);
    });

  });

});