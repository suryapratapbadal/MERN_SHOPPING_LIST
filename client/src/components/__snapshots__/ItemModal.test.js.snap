// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemModal /> render() renders the component 1`] = `
<div>
  <Button
    color="dark"
    onClick={[Function]}
    style={
      Object {
        "marginBottom": "2rem",
      }
    }
    tag="button"
  >
    Add Item
  </Button>
  <Modal
    autoFocus={true}
    backdrop={true}
    backdropTransition={
      Object {
        "mountOnEnter": true,
        "timeout": 150,
      }
    }
    centered={false}
    fade={true}
    isOpen={false}
    keyboard={true}
    modalTransition={
      Object {
        "timeout": 300,
      }
    }
    onClosed={[Function]}
    onOpened={[Function]}
    role="dialog"
    toggle={[Function]}
    zIndex={1050}
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      tag="h5"
      toggle={[Function]}
      wrapTag="div"
    >
      Add to Shopping List
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <Form
        onSubmit={[Function]}
        tag="form"
      >
        <FormGroup
          tag="div"
        >
          <Label
            for="item"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Item
          </Label>
          <Input
            id="item"
            name="name"
            onChange={[Function]}
            placeholder="Add shooping item"
            type="text"
          />
          <Button
            block={true}
            color="dark"
            style={
              Object {
                "marginTop": "2rem",
              }
            }
            tag="button"
          >
            Add Item
          </Button>
        </FormGroup>
      </Form>
    </ModalBody>
  </Modal>
</div>
`;
